dlib/matrix/matrix_fft.h

```diff
@@ -9,837 +9,142 @@
     }
 
 // ----------------------------------------------------------------------------------------
 
-    // Call the MKL DFTI implementation in these cases 
-
-    inline matrix<std::complex<double>,0,1> fft (const matrix<std::complex<double>,0,1>& data)
-    {
-        return call_mkl_fft(data, false);
-    }
-    inline matrix<std::complex<double>,0,1> ifft(const matrix<std::complex<double>,0,1>& data)
-    {
-        return call_mkl_fft(data, true) / data.size();
-    }
-    inline matrix<std::complex<double>,1,0> fft (const matrix<std::complex<double>,1,0>& data)
-    {
-        return call_mkl_fft(data, false);
-    }
-    inline matrix<std::complex<double>,1,0> ifft(const matrix<std::complex<double>,1,0>& data)
-    {
-        return call_mkl_fft(data, true) / data.size();
-    }
-    inline matrix<std::complex<double> > fft (const matrix<std::complex<double> >& data)
-    {
-        return call_mkl_fft(data, false);
-    }
-    inline matrix<std::complex<double> > ifft(const matrix<std::complex<double> >& data)
+    template < typename T, long NR, long NC, typename MM, typename L >
+    void ifft_inplace (matrix<std::complex<T>,NR,NC,MM,L>& data)
     {
-        return call_mkl_fft(data, true) / data.size();
+        kiss_fft({(int)data.nr(), (int)data.nc()}, &data(0,0), &data(0,0), true);
     }
 
-    inline void fft_inplace (matrix<std::complex<double>,0,1>& data)
-    {
-        call_mkl_fft_inplace(data, false);
-    }
-    inline void ifft_inplace(matrix<std::complex<double>,0,1>& data)
-    {
-        call_mkl_fft_inplace(data, true);
-    }
-    inline void fft_inplace (matrix<std::complex<double>,1,0>& data)
-    {
-        call_mkl_fft_inplace(data, false);
-    }
-    inline void ifft_inplace(matrix<std::complex<double>,1,0>& data)
+// ----------------------------------------------------------------------------------------
+    
+    template <typename EXP>
+    matrix<std::complex<typename EXP::type>> fftr (const matrix_exp<EXP>& data)
     {
-        call_mkl_fft_inplace(data, true);
+        // You have to give a complex matrix
+        COMPILE_TIME_ASSERT(std::is_floating_point<typename EXP::type>::value);
+        DLIB_ASSERT(data.nc() % 2 == 0, "last dimension needs to be even otherwise ifftr(fftr(data)) won't have matching dimensions : " << data.nc());
+        matrix<typename EXP::type> in(data);
+        matrix<std::complex<typename EXP::type>> out(in.nr(), in.nc()/2+1);
+        kiss_fftr({(int)in.nr(), (int)in.nc()}, &in(0,0), &out(0,0));
+        return out;
     }
 
-    inline void fft_inplace (matrix<std::complex<double> >& data)
-    {
-        call_mkl_fft_inplace(data, false);
-    }
-    inline void ifft_inplace(matrix<std::complex<double> >& data)
+// ----------------------------------------------------------------------------------------
+    
+    template <typename EXP>
+    matrix<typename remove_complex<typename EXP::type>::type> ifftr (const matrix_exp<EXP>& data)
```

Comment:
I'm not convinced there should be `fftr` and `ifftr`.  The input type of the matrix tells us if it's real or complex, so we shouldn't force the user to tell us that twice by also putting `r` at the end of the function's name.  There can still be overloads specifically for real inputs though.  But they should just be called `fft()` and `ifft()`

Reply:
```ifft``` might be confusing. Both a complex inverse FFT and real inverse FFT require complex input...

Reply:
@davisking Don't you think using ```fftr``` and ```ifftr``` make it clear to the user what the fft is actually doing in this case. It also means we can check input type, shape, etc. A lot of libraries do this, like numpy, torch, etc.

Reply:
> `ifft` might be confusing. Both a complex inverse FFT and real inverse FFT require complex input...

Let me clarify: a forward real FFT is r2c, while a backward real FFT is c2r. A normal FFT is c2c. So both c2r and c2c take complex input. So just having ```ifft``` would be ambiguous.

Reply:
Sorry, I wrote that very unclearly.  Yes, the inverse fft is going to have complex inputs.  But the forward fft can take real inputs just fine.  There isn't any need to make the user write `ffti(whatever)` when we know the type of `whatever`.  There shouldn't be anything to check, as we do the checking by having the appropriate overload.

Then `ifft()` is always a function that maps from a complex input to a complex output.  Someone can just do ```
matrix<double> m = real(ifft(some_complex_matrix));
```
if they want just the real part.  It's possible to make `real(ifft(some_complex_matrix))` map to a different inverse fft function than just `ifft(some_complex_matrix)`.  But it is simpler to write `ifftr` and just document that it's logically equivalent to `real(ifft())`, so it's got that going for it and is maybe fine.  But `fftr` has no such complications.

Reply:
ok i will go with having ```fft``` which can be complex to complex or real to complex, ```ifft``` which is complex to complex only and ```ifftr``` which is complex to real. That's my personal preference. And the API is then crystal clear as to what is happening

Reply:
you make a good point that there should be a unit test that checks ```real(ifft(x)) == ifftr(x)```